lib/pos_blockchain.go

```diff
@@ -100,160 +101,165 @@ func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, v
+		blockNode, innerErr := bc.storeBlockInBlockIndex(block)
+		if innerErr != nil {
+			return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem adding block to block index: %v", innerErr)
 		}
+		return blockNode, nil
+	}
+
+	// At this point, we know the parent block is validation. We can now perform all other validations
+	// on the current block. First we store the block in the block index.
+	blockNode, err = bc.storeBlockInBlockIndex(block)
+	if err != nil {
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem adding block to block index: %v", err)
 	}
 
-	if err = bc.validateDeSoBlockPoS(block); err != nil {
-		// TODO: If validations fail, do we want to even bother storing the full block? Are there any validations that
-		// wouldn't result in us wanting to mark the block as StatusBlockValidateFailed?
-		// Do we ONLY mark it as ValidateFailed if we hit a rule error? We can check this with IsRuleError(err)
-		// If validations fail, we can update the block index with the block status ValidateFailed
-		isRuleError := IsRuleError(err)
-		_ = isRuleError
-		blockNode, storeInBlockIndexErr = bc.storeValidateFailedBlockInBlockIndex(block)
-		if storeInBlockIndexErr != nil {
-			return blockNode, errors.Wrapf(err, "isValidPoSBlock: Problem adding validate failed block to block index: %v", storeInBlockIndexErr)
+	// Check if the block is properly formed and passes all basic validations.
+	if err = bc.isValidBlockPoS(block); err != nil {
+		blockNode, innerErr := bc.storeValidateFailedBlockInBlockIndex(block)
+		if innerErr != nil {
+			return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem adding validate failed block to block index: %v", innerErr)
 		}
-		return blockNode, errors.Wrap(err, "isValidPoSBlock: block validation failed: ")
+		return blockNode, nil
 	}
 
-	// We know the utxoView will be valid because we check that all ancestor blocks have been validated.
+	// We expect the utxoView for the parent block to be valid because we check that all ancestor blocks have
+	// been validated.
 	utxoView, err := bc.getUtxoViewAtBlockHash(*block.Header.PrevBlockHash)
 	if err != nil {
-		// Here we want to at least store this block since we don't know if it's invalid. Hitting
-		// an error here should really never happen.
-		blockNode, storeInBlockIndexErr = bc.storeBlockInBlockIndex(block)
-		if storeInBlockIndexErr != nil {
-			return blockNode, errors.Wrapf(err, "isValidPoSBlock: Problem adding block to block index: %v", storeInBlockIndexErr)
+		// Here we want to at least store this block since we don't know if it's invalid.
+		// Hitting an error here should really never happen.
+		blockNode, innerErr := bc.storeBlockInBlockIndex(block)
+		if innerErr != nil {
+			return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem adding block to block index: %v", innerErr)
 		}
-		return blockNode, errors.Wrap(err, "processBlockPoS: Problem initializing UtxoView: ")
+		return blockNode, nil
```

Comment:
Since we store the block on line 153, we don't need to store it again here. I was thinking the store block on line 153 is basically saying "Okay we have a parent that is validated, we should at least store this block now". This also has the nice property of allowing us to avoid calling `storeBlockInBlockIndex` every time we encounter an error that doesn't result in the block being marked as Validate Failed.

Reply:
I updated it to store the block with validate failed status. In this case, if the UtxoView can't build correctly for the block, then it means the block is bad.